File: deltalognormal_re.cpp
#include <TMB.hpp>
template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(y1_i);
  DATA_VECTOR(y2_i);
  DATA_MATRIX(X1_ij);
  DATA_MATRIX(X2_ij);

  DATA_IVECTOR(factor1_i);
  DATA_IVECTOR(factor2_i);
  DATA_INTEGER(nk1);
  DATA_INTEGER(nk2);

  DATA_MATRIX(X1_pred_ij);
  DATA_MATRIX(X2_pred_ij);

  PARAMETER_VECTOR(b1_j);
  PARAMETER_VECTOR(b2_j);

  PARAMETER_VECTOR(z1_k);
  PARAMETER_VECTOR(z2_k);

  PARAMETER(log_sigma);
  PARAMETER(log_sigma_z1);
  PARAMETER(log_sigma_z2);

  int n1 = y1_i.size();
  int n2 = y2_i.size();

  Type jnll = 0.0; // initialize joint negative log likelihood

  // Linear predictor
  vector<Type> linear_predictor1_i(n1);
  vector<Type> linear_predictor2_i(n2);
  linear_predictor1_i = X1_ij * b1_j;
  linear_predictor2_i = X2_ij * b2_j;

  for(int i = 0; i < n1; i++){
    jnll -= dbinom_robust(y1_i(i), Type(1.0),
      linear_predictor1_i(i) + z1_k(factor1_i(i)), true);
  }
  for(int i = 0; i < n2; i++){
    jnll -= dnorm(y2_i(i),
      linear_predictor2_i(i) + z2_k(factor2_i(i)), exp(log_sigma), true);
  }

  // Probability of random coefficients
  for(int k = 0; k<nk1; k++){
    jnll -= dnorm(z1_k(k), Type(0.0), exp(log_sigma_z1), true);
  }
  for(int k = 0; k<nk2; k++){
    jnll -= dnorm(z2_k(k), Type(0.0), exp(log_sigma_z2), true);
  }

  Type sigma = exp(log_sigma);
  REPORT(sigma);

  vector<Type> linear_prediction1_i(n1);
  vector<Type> linear_prediction2_i(n2);
  linear_prediction1_i = X1_pred_ij * b1_j;
  linear_prediction2_i = X2_pred_ij * b2_j;

  vector<Type> prediction =
    exp(linear_prediction1_i) / (exp(linear_prediction1_i) + 1) *
    exp(linear_prediction2_i);

  vector<Type> log_prediction = log(prediction);
  REPORT(prediction);
  ADREPORT(prediction);
  REPORT(log_prediction);
  ADREPORT(log_prediction);

  return jnll;
}

File: deltalognormal_re.R
# library(TMB)
# compile("deltalognormal_re.cpp")
# dyn.load(dynlib("deltalognormal_re"))
#
# f1 <- pos_catch ~ year_factor
# f2 <- log(spp_catch) ~ year_factor
# mm1 <- model.matrix(f1, data = d_retained)
# mm2 <- model.matrix(f2, data = subset(d_retained, pos_catch == 1))
#
# get_most_freq_factor <- function(x) {
#   names(rev(sort(table(x))))[1]
# }
# newdata <- expand.grid(
#   year_factor = sort(unique(d_retained$year_factor)),
#   month_factor = get_most_freq_factor(d_retained$month_factor),
#   depth_band = get_most_freq_factor(d_retained$depth_band),
#   latitude_band = get_most_freq_factor(d_retained$latitude_band),
#   dfo_locality = get_most_freq_factor(d_retained$locality_code),
#   vessel_name = get_most_freq_factor(d_retained$vessel_name),
#   hours_fished = mean(d_retained$hours_fished),
#   pos_catch = NA,
#   spp_catch = NA,
#   best_depth = mean(d_retained$best_depth))
# mm_pred <- mm2[1:length(unique(d_retained$year)), ]
# for (i in 1:ncol(mm_pred)) {
#   for (j in 1:nrow(mm_pred)) {
#     mm_pred[j, i] <- 0
#   }}
# mm_pred[,1] <- 1
# for (i in 1:ncol(mm_pred)) {
#   for (j in 1:nrow(mm_pred)) {
#     if (i == j)
#       mm_pred[j, i] <- 1
#   }}
#
# m_bin <- speedglm::speedglm(f1, data = d_retained, family = binomial(link = "logit"))
# m_pos <- lm(f2, data = subset(d_retained, pos_catch == 1))
#
# bin_vess <- as.numeric(as.factor(as.character(d_retained$vessel_name)))
# pos_vess <- as.numeric(as.factor(as.character(subset(d_retained, pos_catch == 1)$vessel_name)))

obj <- MakeADFun(
  data = list(
    X1_ij = mm1, y1_i = d_retained$pos_catch,
    X2_ij = mm2, y2_i = log(subset(d_retained, pos_catch == 1)$spp_catch),

    factor1_k = bin_vess,
    factor2_k = pos_vess,

    X1_pred_ij = mm_pred,
    X2_pred_ij = mm_pred,

    nk1 = length(unique(bin_vess)),
    nk2 = length(unique(pos_vess))),
  # parameters = list(b1_j = rep(0, ncol(mm1)), b2_j = rep(0, ncol(mm2)),
  parameters = list(
    b1_j = coef(m_bin) + rnorm(length(coef(m_bin)), 0, 0.05),
    b2_j = coef(m_pos)  + rnorm(length(coef(m_bin)), 0, 0.05),

    z1_k = rep(0, length(unique(bin_vess))),
    z2_k = rep(0, length(unique(pos_vess))),

    log_sigma = log(1),
    log_sigma_z1 = log(1),
    log_sigma_z2 = log(1)),
  # random = c("z1_k", "z2_k"),
  DLL = "deltalognormal_re")
